### PR TITLE
feat(lumin-pdf): add Lumin PDF eSignature integration

### PR DESCRIPTION
**Pull request by Shubhank-Jonnada**
## Summary

- Adds a new **Lumin PDF** integration -- a cloud-based PDF editing and eSignature platform
- Implements **15 actions** covering user/workspace management, template discovery, the full signature request lifecycle, document upload, agreement management, and reminders
- Uses **custom API key** authentication (`X-API-KEY` header)
- Targets **`autohive-integrations-sdk ~=2.0.0`** with `FetchResponse`, `ActionError`, and `Integration.load()` patterns

## Actions

| Action | Description |
|--------|-------------|
| `get_current_user` | Get authenticated user info |
| `get_workspace` | Get current workspace details |
| `list_workspace_members` | List all workspace members |
| `list_templates` | List all templates |
| `get_template` | Get a specific template by ID |
| `send_signature_request` | Send a signature request with one or more PDFs |
| `get_signature_request` | Get status and details of a signature request |
| `generate_signing_link` | Generate an embedded signing URL for a signer |
| `send_reminder` | Send reminder emails to pending signers |
| `cancel_signature_request` | Cancel a pending signature request |
| `update_signature_request` | Update expiry date on a signature request |
| `download_signed_document` | Download the signed document as a file URL |
| `upload_document` | Upload a document from URL or template |
| `generate_document_from_template` | Generate a filled document from a template |
| `create_agreement` | Create an agreement from a Lumin template |
| `download_agreement` | Download a completed agreement as a file URL |
| `send_from_template` | Send a signature request directly from a template |

### Unit tests (46 passed)

All actions covered with mocked `FetchResponse` — success paths, error paths, and edge cases (e.g. `email` vs `email_address` normalisation, `file_url` vs `file_urls`, list vs dict response shapes for members/templates).

Tests follow repo-standard patterns: use `make_context` from root conftest, no `@pytest.mark.asyncio` decorators, no custom fixtures.

```
46 passed in 0.53s
```

### Live integration tests

Run against the real Lumin PDF API. Tests requiring Lumin-type templates (send_from_template, create_agreement, download_agreement) are skipped in accounts without them — these are pending live verification once templates are available in the test workspace.

```
PASSED  TestGetCurrentUser::test_returns_user_info
PASSED  TestGetWorkspace::test_returns_workspace
PASSED  TestListWorkspaceMembers::test_returns_members
PASSED  TestListTemplates::test_returns_templates
PASSED  TestSignatureRequestLifecycle::test_full_lifecycle  ← send → get → signing link → cancel
PASSED  TestUploadDocument::test_upload_from_url
PASSED  TestUpdateSignatureRequest::test_extends_expiry
PASSED  TestSendReminder::test_sends_reminder
SKIPPED TestGetTemplate — no templates in workspace
SKIPPED TestSendFromTemplate — no lumin-type templates in workspace
SKIPPED TestCreateAgreement — no lumin-type templates in workspace
SKIPPED TestDownloadAgreement — no lumin-type templates in workspace
SKIPPED TestGenerateDocumentFromTemplate — no lumin-type templates in workspace
```

The skipped tests will run once lumin-type templates exist in the sandbox — the test logic is implemented and will execute automatically when the workspace is ready.

## Validation

```
validate_integration.py  — 0 errors, 0 warnings
check_code.py            — syntax, imports, JSON, lint, format, security, deps, config-code sync all passed
```

### CODE REVIEW

**Comment by Shubhank-Jonnada** [comment]
All 10 actions unit tested (26 tests) + live tested against the real API, 5 passed, 1 skipped (get_template, no templates in the account).

@TheRealAgentK @NinosMan

File: lumin-pdf/lumin_pdf.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+                "title": inputs["title"],
+                "signers": signers,
+            }
+            if inputs.get("file_urls"):
+                body["file_urls"] = inputs["file_urls"]
+            elif inputs.get("file_url"):
+                body["file_url"] = inputs["file_url"]
+            if inputs.get("message"):
+                body["message"] = inputs["message"]
+            body["expires_at"] = inputs.get("due_date") or int((time.time() + 30 * 86400) * 1000)
```
[the remark this answers is not in the document]
Comment (Shubhank-Jonnada) [reply lumin-pdf/lumin_pdf.py]: True, fixed. Now converts ISO 8601 → epoch milliseconds before sending to Lumin.

File: lumin-pdf/lumin_pdf.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    async def execute(self, inputs: Dict[str, Any], context: ExecutionContext) -> ActionResult:
+        try:
+            params: Dict[str, Any] = {"page": inputs.get("page", 1)}
+            if inputs.get("limit") is not None:
+                params["limit"] = inputs["limit"]
+            response = await context.fetch(
+                f"{BASE_URL}/templates",
+                method="GET",
+                headers=_auth_headers(context),
+                params=params,
```
[the remark this answers is not in the document]
Comment (Shubhank-Jonnada) [reply lumin-pdf/lumin_pdf.py]: False

File: lumin-pdf/lumin_pdf.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+
+@lumin_pdf.action("send_reminder")
+class SendReminderAction(ActionHandler):
+    async def execute(self, inputs: Dict[str, Any], context: ExecutionContext) -> ActionResult:
+        try:
+            req_id = inputs["signature_request_id"]
+            body: Dict[str, Any] = {}
+            if inputs.get("emails"):
+                body["emails"] = inputs["emails"]
```
[the remark this answers is not in the document]
Comment (Shubhank-Jonnada) [reply lumin-pdf/lumin_pdf.py]: False

File: lumin-pdf/lumin_pdf.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+            response = await context.fetch(
+                f"{BASE_URL}/signature_request/{req_id}/file",
+                method="GET",
+                headers=_auth_headers(context),
+                params={"type": doc_type},
+            )
+            _raise_for_status(response)
+            data = response.data
+            file_url = (
+                data.get("file_url") or data.get("url") or data.get("download_url", "")
```
Comment (TheRealAgentK) [inline lumin-pdf/lumin_pdf.py]: The download actions need a response-shape fix. Lumin’s current file endpoints return `signed_url`, but both download actions only look for `file_url`, `url`, or `download_url`.

That means a successful API call can still return `file_url: ""` to Autohive.

Can we include `signed_url` in the fallback chain, ideally first, and update the unit mocks to use the documented response shape?

File: lumin-pdf/lumin_pdf.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+            elif inputs.get("file_url"):
+                body["file_url"] = inputs["file_url"]
+            if inputs.get("message"):
+                body["message"] = inputs["message"]
+            due_date = inputs.get("due_date")
+            if due_date:
+                from datetime import datetime, timezone
+
+                body["expires_at"] = int(
+                    datetime.fromisoformat(due_date).replace(tzinfo=timezone.utc).timestamp() * 1000
```
Comment (TheRealAgentK) [inline lumin-pdf/lumin_pdf.py]: The due date handling has a timezone bug. `replace(tzinfo=timezone.utc)` overwrites the timezone instead of converting it.

For example, `2026-12-31T00:00:00-08:00` should convert to `2026-12-31T08:00:00Z`, but replacing the timezone makes it `2026-12-31T00:00:00Z`.

Can we move this into a helper that only assumes UTC when the parsed datetime is naive, then reuse it in the three due-date paths?

File: lumin-pdf/lumin_pdf.py [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+            if not emails:
+                # fetch pending signers to remind all
+                sr = await context.fetch(
+                    f"{BASE_URL}/signature_request/{req_id}",
+                    method="GET",
+                    headers=_auth_headers(context),
+                )
+                _raise_for_status(sr)
+                signers = sr.data.get("signature_request", sr.data).get("signers", [])
+                emails = [s["email"] for s in signers if s.get("status") == "NEED_TO_SIGN" and s.get("email")]
```
Comment (TheRealAgentK) [inline lumin-pdf/lumin_pdf.py]: The optional `emails` path looks like it reads the wrong signer field. The code uses `s["email"]`, but Lumin signer objects use `email_address`.

So if users omit `emails`, this can collect an empty list and post `{"emails": []}` instead of reminding pending signers.

Can we read `email_address` with an `email` fallback, and return a helpful error if no pending signer emails are found?

File: lumin-pdf/tests/test_lumin_pdf_unit.py [hunk 7]
```diff
@@ -0,0 +1,620 @@
...
+        )
+
+        assert result.type == ResultType.ACTION_ERROR
+
+
+class TestDownloadSignedDocument:
+    async def test_returns_file_url(self, make_context):
+        ctx = make_context(auth={"api_key": API_KEY})
+        ctx.fetch.return_value = FetchResponse(
+            status=200, headers={}, data={"file_url": "https://cdn.lumin.com/signed.pdf"}
```
Comment (TheRealAgentK) [inline lumin-pdf/tests/test_lumin_pdf_unit.py]: The unit coverage is broad, but a few mocks mirror the implementation rather than Lumin’s documented API shapes, which lets these issues slip through:

- download mocks should return `signed_url`
- reminder signer mocks should use `email_address`
- due date tests should assert a known epoch value, including an offset-aware ISO string

That should make the tests catch contract regressions without changing the overall coverage shape.

**Review by TheRealAgentK — COMMENTED** [review]
Thanks for putting this integration together. I left a few focused comments around Lumin API response shapes, reminder email discovery, timezone handling, and making the unit mocks catch those contracts.

File: lumin-pdf/lumin_pdf.py (hunk 4 above)
Comment (Shubhank-Jonnada) [reply lumin-pdf/lumin_pdf.py]: Fixed in https://github.com/Autohive-AI/autohive-integrations/commit/bfe5c23 - signed_url added first in the fallback chain for both download actions, mocks updated to match.

File: lumin-pdf/lumin_pdf.py (hunk 5 above)
Comment (Shubhank-Jonnada) [reply lumin-pdf/lumin_pdf.py]: Fixed in https://github.com/Autohive-AI/autohive-integrations/commit/84558d1 - extracted _to_epoch_millis helper that only assumes UTC for naive datetimes. Unit tests now assert exact epoch values including an offset-aware ISO string.

File: lumin-pdf/lumin_pdf.py (hunk 6 above)
Comment (Shubhank-Jonnada) [reply lumin-pdf/lumin_pdf.py]: Fixed in https://github.com/Autohive-AI/autohive-integrations/commit/b56d554 - switched to email_address with email fallback, returns an error if no pending signers are found.

File: lumin-pdf/tests/test_lumin_pdf_unit.py (hunk 7 above)
Comment (Shubhank-Jonnada) [reply lumin-pdf/tests/test_lumin_pdf_unit.py]: covered across the 3 commits above, download mocks use signed_url, reminder mocks use email_address, and due date tests now assert exact epoch values including an offset-aware ISO string.

**Comment by Shubhank-Jonnada** [comment]
Thanks for the review @TheRealAgentK  ! All points addressed, ready for a re-review

**Review by TheRealAgentK — APPROVED** [review]
LGTM
